source/common/stats/stats_impl.h

```diff
@@ -373,30 +384,60 @@ class SourceImpl : public Source {
 };
 
 /**
- * Implementation of RawStatDataAllocator that uses an unordered set to store
- * RawStatData pointers.
+ * This structure is an alternate backing store for both CounterImpl and GaugeImpl. It is designed
+ * so that it can be allocated efficiently from the heap on demand.
+ */
+struct HeapStatData {
+  explicit HeapStatData(absl::string_view key);
+
+  /**
+   * @returns absl::string_view the name as a string_view.
+   */
+  absl::string_view key() const { return name_; }
+
+  std::atomic<uint64_t> value_{0};
+  std::atomic<uint64_t> pending_increment_{0};
+  std::atomic<uint16_t> flags_{0};
+  std::atomic<uint16_t> ref_count_{1};
+  std::string name_;
+};
+
+/**
+ * Implementation of StatDataAllocator using a pure heap-based strategy, so that
+ * Envoy implementations that do not require hot-restart can use less memory.
  */
-class HeapRawStatDataAllocator : public RawStatDataAllocator {
+class HeapStatDataAllocator : public StatDataAllocatorImpl<HeapStatData> {
 public:
-  // RawStatDataAllocator
-  ~HeapRawStatDataAllocator() { ASSERT(stats_.empty()); }
-  RawStatData* alloc(const std::string& name) override;
-  void free(RawStatData& data) override;
+  explicit HeapStatDataAllocator(const StatsOptions& options) : StatDataAllocatorImpl(options) {}
+  // HeapStatDataAllocator() {}
+  ~HeapStatDataAllocator() { ASSERT(stats_.empty()); }
+
+  // StatDataAllocator
+  HeapStatData* alloc(const std::string& name) override;
+  void free(HeapStatData& data) override;
 
 private:
-  struct RawStatDataHash_ {
-    size_t operator()(const RawStatData* a) const { return HashUtil::xxHash64(a->key()); }
+  struct HeapStatHash_ {
+    size_t operator()(const HeapStatData* a) const { return HashUtil::xxHash64(a->key()); }
   };
-  struct RawStatDataCompare_ {
-    bool operator()(const RawStatData* a, const RawStatData* b) const {
+  struct HeapStatCompare_ {
+    bool operator()(const HeapStatData* a, const HeapStatData* b) const {
       return (a->key() == b->key());
     }
   };
-  typedef std::unordered_set<RawStatData*, RawStatDataHash_, RawStatDataCompare_> StringRawDataSet;
 
-  // An unordered set of RawStatData pointers which keys off the key()
+  // TODO(jmarantz): storing the set of stats as a string-set wastes an
+  // enourmous amount of of memory, because all the stats are composed using the
+  // same set of cluster names and patterns found in
+  // source/common/config/well_known_names.cc. This should eventually be changed to
+  // store a symbol-table of decomposed stat segments (e.g. split on "." to start) and
+  // arrays referencing those segments. The segments can be ref-counted so they
+  // no longer consume memory once deleted.
```

Comment:
Either remove TODO or point to @ambuc 's latest PR here.

Reply:
Rewrote to reference the new PR.